src/core/linear_solver/src/preconditioner/4C_linear_solver_preconditioner_muelu.cpp

```diff
@@ -158,308 +164,42 @@ void Core::LinearSolver::MueLuPreconditioner::setup(bool create, Epetra_Operator
-    Teuchos::rcp_dynamic_cast<const Xpetra::StridedMap<LO, GO, NO>>(
-        myA21->getRowMap("stridedMaps"), true);
-    Teuchos::rcp_dynamic_cast<const Xpetra::StridedMap<LO, GO, NO>>(
-        myA21->getColMap("stridedMaps"), true);
-  }
-
-  // Re-create or re-use the preconditioner?
-  if (create)
-  {
-    // (Re-)create the preconditioner
-
-    // free old matrix first
-    P_ = Teuchos::null;
-
-    if (!contactList.isParameter("MUELU_XML_FILE"))
-      FOUR_C_THROW(
-          "XML-file w/ MueLu preconditioner configuration is missing in solver parameter list. "
-          "Please set it as entry 'MUELU_XML_FILE'.");
-    std::string xml_file = contactList.get<std::string>("MUELU_XML_FILE");
-
-    MueLu::ParameterList mueluParams;
-    Teuchos::updateParametersFromXmlFileAndBroadcast(
-        xml_file, Teuchos::Ptr<MueLu::ParameterList>(&mueluParams), *comm);
-
-    // Get/compute nullspace vectors
-    Teuchos::RCP<Xpetra::MultiVector<SC, LO, GO, NO>> nullspace11 = Teuchos::null;
-    Teuchos::RCP<Xpetra::MultiVector<SC, LO, GO, NO>> nullspace22 = Teuchos::null;
-    {
-      // Extract pre-computed nullspace for block (0,0) from 4C's MueLu parameter list
-      nullspace11 = Core::LinearSolver::Parameters::extract_nullspace_from_parameterlist(
-          *stridedRangeMapPrimal, contactList);
-
-      // Compute default nullspace for block (1,1)
-      {
-        const int dimNS2 = numdf;
-        nullspace22 =
-            Xpetra::MultiVectorFactory<SC, LO, GO, NO>::Build(stridedRangeMapDual, dimNS2);
-
-        for (int i = 0; i < dimNS2; ++i)
-        {
-          Teuchos::ArrayRCP<Scalar> nsValues22 = nullspace22->getDataNonConst(i);
-          int numBlocks = nsValues22.size() / dimNS2;
-          for (int j = 0; j < numBlocks; ++j)
-          {
-            nsValues22[j * dimNS2 + i] = 1.0;
-          }
-        }
+        H_->GetLevel(0)->Set("Nullspace" + std::to_string(block + 1), nullspace);
       }
-    }
-
-    // ToDo (mayr.mt) Switch to CreateXpetraPreconditioner. Pass nullspace via "user data" sublist
-    // and use xml-entry "Fine level nullspace" in NullspaceFactory.
-    //
 
-    // ParameterListInterpreter mueLuFactory(xml_file, *comm);
-    MueLu::ParameterListInterpreter<SC, LO, GO, NO> mueLuFactory(mueluParams, comm);
-
-    Teuchos::RCP<MueLu::Hierarchy<SC, LO, GO, NO>> H = mueLuFactory.CreateHierarchy();
-    H->GetLevel(0)->Set("A", Teuchos::rcp_dynamic_cast<Xpetra::Matrix<SC, LO, GO, NO>>(bOp));
-    H->GetLevel(0)->Set("Nullspace1", nullspace11);
-    H->GetLevel(0)->Set("Nullspace2", nullspace22);
-    H->GetLevel(0)->Set("Primal interface DOF map",
-        Teuchos::rcp_dynamic_cast<const Xpetra::Map<LO, GO, NO>>(xSlaveDofMap, true));
+      if (muelulist_.sublist("Belos Parameters").isParameter("contact slaveDofMap"))
+      {
+        Teuchos::RCP<Epetra_Map> ep_slave_dof_map =
+            muelulist_.sublist("Belos Parameters")
```

Comment:
I'm not super happy about the fact, that a map used in a preconditioner is stored in the linear solver parameterlist, but not urgent for now.